nvtabular/io/shuffle.py

```diff
@@ -47,10 +49,20 @@ def _check_shuffle_arg(shuffle):
     return shuffle
 
 
-def _shuffle_gdf(gdf, gdf_size=None):
-    """Shuffles a cudf dataframe, returning a new dataframe with randomly
+def _shuffle_df(df, size=None):
+    """Shuffles a DataFrame, returning a new dataframe with randomly
     ordered rows"""
-    gdf_size = gdf_size or len(gdf)
-    arr = cp.arange(gdf_size)
-    cp.random.shuffle(arr)
-    return gdf.iloc[arr]
+    size = size or len(df)
+    # NOTE: We can use np.arange for both gpu and cpu-backed
+    # dataframes once NEP-35 is fully accepted (`like` argument).
+    # This should be available in numpy>=1.20
+    if isinstance(df, pd.DataFrame):
+        arr = np.arange(size)
+        np.random.shuffle(arr)
+    else:
+        arr = cp.arange(size)
+        # Note that np.random.shuffle "should" Work for both gpu
+        # and cpu (via NEP-18), but it seems the cupy API is
+        # still needed here for correct behavior.
+        cp.random.shuffle(arr)
```

Comment:
Yeah - it seems like this should work, but fails horribly.

I just tried it out using np.random.shuffle on a cupy array - and while it didn't throw an exception it mangled the results pretty badly:

```python
arr = cupy.arange(10)
# arr = array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

np.random.shuffle(arr)
# arr = array([0, 1, 1, 3, 1, 5, 6, 3, 1, 2]) 
```

I don't think this is a nep-18 problem though - I think its a legit bug in cupy.  For non-numpy arrays, numpy shuffle just swaps entries:

https://github.com/numpy/numpy/blob/9f12028b9453c17b72b26355fd503e512af96a5d/numpy/random/_generator.pyx#L4455-L4458

We can get the same bug happening here without involving numpy at all


``` python
x = cupy.arange(10)
# x = array([0., 1., 9., 3., 4., 5., 6., 7., 8., 9.])

# swap the first last entries
x[0], x[9] = x[9], x[0]
# x = array([9, 1, 2, 3, 4, 5, 6, 7, 8, 9])
```

This seems to be because of https://github.com/cupy/cupy/issues/2824

Reply:
Whoa - good find! I added a note about the cupy bug